test/e2e/scheduling/nvidia-gpus.go

```diff
@@ -189,3 +192,101 @@ var _ = SIGDescribe("[Feature:GPUDevicePlugin]", func() {
 		testNvidiaGPUs(f)
 	})
 })
+
+func testNvidiaGPUsJob(f *framework.Framework) {
+	_ = SetupNVIDIAGPUNode(f, false)
+	// Job set to have 5 completions with parallelism of 1 to ensure that it lasts long enough to experience the node recreation
+	completions := int32(5)
+	By("Starting GPU job")
+	StartJob(f, completions)
+
+	job, err := jobutil.GetJob(f.ClientSet, f.Namespace.Name, "cuda-add")
+	Expect(err).NotTo(HaveOccurred())
+
+	// make sure job is running by waiting for its first pod to start running
+	err = jobutil.WaitForAllJobPodsRunning(f.ClientSet, f.Namespace.Name, job.Name, 1)
+	Expect(err).NotTo(HaveOccurred())
+
+	numNodes, err := framework.NumberOfRegisteredNodes(f.ClientSet)
+	Expect(err).NotTo(HaveOccurred())
+	nodes, err := framework.CheckNodesReady(f.ClientSet, numNodes, framework.NodeReadyInitialTimeout)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Recreating nodes")
+	gce.RecreateNodes(f.ClientSet, nodes)
+	By("Done recreating nodes")
+
+	By("Waiting for gpu job to finish")
+	err = jobutil.WaitForJobFinish(f.ClientSet, f.Namespace.Name, job.Name)
+	Expect(err).NotTo(HaveOccurred())
+	By("Done with gpu job")
+
+	Expect(job.Status.Failed).To(BeZero(), "Job pods failed during node recreation: %v", job.Status.Failed)
+
+	VerifyJobNCompletions(f, completions)
+}
+
+// Starts a simple CUDA job that requests gpu and the specified number of completions
+func StartJob(f *framework.Framework, completions int32) {
+	var activeSeconds int64 = 3600
+	testJob := jobutil.NewTestJob("succeed", "cuda-add", v1.RestartPolicyAlways, 1, completions, &activeSeconds, 6)
+	testJob.Spec.Template.Spec = v1.PodSpec{
+		RestartPolicy: v1.RestartPolicyOnFailure,
+		Containers: []v1.Container{
+			{
+				Name:    "vector-addition",
+				Image:   imageutils.GetE2EImage(imageutils.CudaVectorAdd),
+				Command: []string{"/bin/sh", "-c", "./vectorAdd && sleep 60"},
+				Resources: v1.ResourceRequirements{
+					Limits: v1.ResourceList{
+						gpuResourceName: *resource.NewQuantity(1, resource.DecimalSI),
+					},
+				},
+			},
+		},
+	}
+	ns := f.Namespace.Name
+	_, err := jobutil.CreateJob(f.ClientSet, ns, testJob)
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("Created job %v", testJob)
+}
+
+// Verifies that the job has completions number of successful pods
+func VerifyJobNCompletions(f *framework.Framework, completions int32) {
+	ns := f.Namespace.Name
+	pods, err := jobutil.GetJobPods(f.ClientSet, f.Namespace.Name, "cuda-add")
+	Expect(err).NotTo(HaveOccurred())
+	createdPods := pods.Items
+	createdPodNames := podNames(createdPods)
+	framework.Logf("Got the following pods for job cuda-add: %v", createdPodNames)
+
+	successes := int32(0)
+	for _, podName := range createdPodNames {
+		f.PodClient().WaitForFinish(podName, 5*time.Minute)
+		logs, err := framework.GetPodLogs(f.ClientSet, ns, podName, "vector-addition")
+		framework.ExpectNoError(err, "Should be able to get logs for pod %v", podName)
+		regex := regexp.MustCompile("PASSED")
```

Comment:
This seems pretty fragile. What does "cuda-vector-add" actually run?

Reply:
cuda-vector-add points to this image: https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go#L206
and runs a cuda program the same as this: https://github.com/tpn/cuda-samples/blob/master/v8.0/0_Simple/vectorAdd/vectorAdd.cu

Reply:
Looks like this won't catch if you fail to free devices:

https://github.com/tpn/cuda-samples/blob/master/v8.0/0_Simple/vectorAdd/vectorAdd.cu#L168-L188

Reply:
I guess that's caught by the `Expect(job.Status.Failed).To(BeZero())` check

Reply:
Yea, that should catch, but do you think this should check for "DONE" instead of "PASSED" to ensure that this device freeing is caught here? The original purpose of the job was to ensure that a cuda pod requiring the nvidia driver could run, which is why I assumed the original job checks for PASSED.

Reply:
This seems fine.

Reply:
This seems fine.

Reply:
Thanks for going through my PR!